### DISCUSSION TITLE
Scalable Graph (size, and possibly memory)

### DISCUSSION THREAD

**Discussion by aFineBoi** [networkx/networkx#6439] [Algorithms]
ML including GNN ones are bottlenecked by memory problems. Consider adding support for multi-level addressing of nodes? 

It's comparable to SSD mapping:
> The naive approach is to use a page-level mapping to map any logical page from the host to a physical page. This mapping policy offers a lot of flexibility, but the major drawback is that the mapping table requires a lot of RAM, which can significantly increase the manufacturing costs. A solution to that would be to map blocks instead of pages, using a block-level mapping. Let’s assume that an SSD drive has 256 pages per block. This means that block-level mapping requires 256 times less memory than page-level mapping, which is a huge improvement for space utilization.
https://codecapsule.com/2014/02/12/coding-for-ssds-part-3-pages-blocks-and-the-flash-translation-layer/

Currently native networkx doesn't support this, because it doesn't support adding edges to nodes that is in different graphs:

![image](https://ars.els-cdn.com/content/image/1-s2.0-S0020025521012469-gr7.jpg)
credit: https://www.sciencedirect.com/science/article/pii/S0020025521012469

**Comment by daStretchyCat** [comment]
no, it's not supported

```
import networkx as nx

G = [nx.Graph() for i in range(10)]

for g in G:
    for i in range(10):
        for j in range(10):
            if i != j:
                g.add_edge(i, j)

# wiring
for i in range(len(G)):
    for j in range(len(G)):
        if i != j:
            G[i].add_edge(0, G[j].nodes()[0])

for g in G:
    nx.draw(g, with_labels=True)

```

**Reply by dschult** [reply]
This is kind of supported actually -- Nodes in NetworkX can be any hashable object.
And we intentionally constructed them so that a graph can be a node in another graph.
The drawing capability doesn't automatically show the graphs nicely as what you show
but there might be some tricky way to generate images for each outer-graph node 
using `draw` on the inner-graphs.

But for many types of graph analysis the graph-as-a-node approach should work. 
For example, the shortest_path suite of weighted algorithms allow arbitrary functions 
to determine the edge weights. Those functions can delve into the inner-graphs 
which are nodes in the outer graph and find the edge weight based on properties 
of the inner-graph.

The following code would create the graph you have a drawing of.
```G = nx.complete_graph([nx.complete_graph(i) for i in range(4)])```
